crates/e2e/tests/e2e/tracking_insufficient_funds.rs

```diff
@@ -65,19 +92,81 @@ async fn test(web3: Web3) {
     );
 
-    let order_is_invalid = || async {
-        let events = crate::database::events_of_order(services.db(), &uid).await;
-        events.last().map(|e| e.label) == Some(OrderEventLabel::Invalid)
+    let orders_are_invalid = || async {
+        let events_a = crate::database::events_of_order(services.db(), &uid_a).await;
+        let events_b = crate::database::events_of_order(services.db(), &uid_b).await;
+        let order_a_correct_events = events_a.into_iter().map(|e| e.label).collect::<Vec<_>>()
+            == vec![OrderEventLabel::Created, OrderEventLabel::Invalid];
+        let order_b_correct_events = events_b.into_iter().map(|e| e.label).collect::<Vec<_>>()
+            == vec![OrderEventLabel::Created, OrderEventLabel::Invalid];
+        order_a_correct_events && order_b_correct_events
+    };
+    wait_for_condition(TIMEOUT, orders_are_invalid)
+        .await
+        .unwrap();
+
+    // Make sure that the next update is happened and no new Invalid event is
+    // received for the `order_b`. `order_a` is required to track if the next update
+    // is happened.
+    tx_value!(
+        trader_a.account(),
+        to_wei(3),
+        onchain.contracts().weth.deposit()
+    );
+    onchain.mint_block().await;
+    let orders_updated = || async {
+        let events_a = crate::database::events_of_order(services.db(), &uid_a).await;
+        let events_b = crate::database::events_of_order(services.db(), &uid_b).await;
+        let order_b_correct_events = events_b.into_iter().map(|e| e.label).collect::<Vec<_>>()
+            == vec![OrderEventLabel::Created, OrderEventLabel::Invalid];
+        events_a.len() > 2
+            && check_non_consecutive_invalid_events(&events_a)
+            && order_b_correct_events
+    };
+    wait_for_condition(TIMEOUT, orders_updated).await.unwrap();
+
+    // Another update should proceed with the order_b.
+    tx_value!(
+        trader_b.account(),
+        to_wei(3),
+        onchain.contracts().weth.deposit()
+    );
+    onchain.mint_block().await;
+    let orders_updated = || async {
+        let events_a = crate::database::events_of_order(services.db(), &uid_b).await;
+        let events_b = crate::database::events_of_order(services.db(), &uid_b).await;
+        events_a.last().map(|o| o.label) == Some(OrderEventLabel::Traded)
+            && events_b.last().map(|o| o.label) == Some(OrderEventLabel::Traded)
+            && check_non_consecutive_invalid_events(&events_a)
+            && check_non_consecutive_invalid_events(&events_b)
     };
-    wait_for_condition(TIMEOUT, order_is_invalid).await.unwrap();
+    wait_for_condition(TIMEOUT, orders_updated).await.unwrap();
+}
+
+fn check_non_consecutive_invalid_events(events: &[OrderEvent]) -> bool {
+    events
+        .iter()
+        .enumerate()
+        .filter_map(|(idx, e)| match e.label {
+            OrderEventLabel::Invalid => Some(idx),
+            _ => None,
+        })
```

Comment:
nit: Slightly verbose 😅 
```suggestion
    !events
        .windows(2)
        .map(|w| (w[0].label, w[1].label))
        .any(|window| window == (OrderEventLabel::Invalid, OrderEventLabel::Invalid))
```